### ISSUE TITLE
Taking one further step of whitebox approach

### ISSUE THREAD

**Issue by ngkel** [Ma-Lab-Berkeley/CRATE#9]
Thank you for open-sourcing such a great work! I have some questions regarding the potential of taking one more step to make the entire CRATE white-box.

As mentioned in the paper, CRATE is trained in a supervised manner with cross-entropy loss to update the dictionary and subspaces parameter, which somehow makes the learning more task-dependent than ReduNet. Is it theoretically rigorous to say that, I can flatten the Z matrix, whose N columns z are learnt representation of tokens with underlying distribution being mixture of Gaussian, then input them to ReduNet and obtain compressed and discriminated representation of the image as a whole? Though I am not sure how the dictionary for tokens should be updated in this case.

**Comment by DruvPai** [comment]
Thanks for the question!
It's true that both CRATE and ReduNet are white-box deep networks, yet their representations are at different labels. As you correctly identified, CRATE operates on the level of patches, and ReduNet on the level of whole images. 
There is nothing saying you _can't_ do what you are suggesting -- after all, the operations are valid in PyTorch -- but in the end you would collapse the whole diverse set of patch representations into a single vector. This is not the most natural operation, nor is it particularly theoretically well-founded -- to faithfully represent a Gaussian mixture for each patch representation, you would need combinatorially many Gaussian mixtures (e.g. classes) for ReduNet to work with without losing information. This is, computationally, horrifically infeasible, so you will lose some information in the representations for what seems like no good reason.
